docs/source/tutorials/websockets.rst

```diff
@@ -0,0 +1,1074 @@
+assigned to ``app.websocket_api.session``.
+This is needed because in order to send websocket responses to API Gateway we
+need to construct a boto3 client. Chalice does not take a direct dependency
+on boto3 or botocore, so we need to provide the Session ourselves.
+
+.. code-block:: python
+
+   from boto3.session import Session
+   app.websocket_api.session = Session()
+
+
+Next we enable the experimental feature ``WEBSOCKETS``. As noted at the top
+of this section, websockets are an experimental feature and are subject to
+API changes.
+
+.. code-block:: python
+
+   app.experimental_feature_flags.update([
+       'WEBSOCKETS'
+   ])
+
+
+To register a websocket handler, and cause Chalice to deploy an
+API Gateway Websocket API we use the ``app.on_ws_message()`` decorator.
+The event parameter here is a wrapper object with some convenience
+parameters attached. The most useful are ``event.connection_id`` and
+``event.body``. The ``connection_id`` is an API Gateway specific identifier
+that allows you to refer to the connection that sent the message. The ``body``
+is the content of the message.
+
+.. code-block:: python
+
+   @app.on_ws_message()
+   def message(event):
+
+
+Since this is an echo server, the message handler simply reads the content it
+received on the socket, and rewrites it back to the same socket. To send a
+message to a socket we call ``app.websocket_api.send(connection_id, message)``.
+In this case, we just use the same ``connection_id`` we got the message from,
+and use the ``body`` we got from the event as the ``message`` to send.
+
+.. code-block:: python
+
+   app.websocket_api.send(
+       connection_id=event.connection_id,
+       message=event.body,
+    )
+
+
+Finally, we catch the exception ``WebsocketDisconnectedError`` which is raised
+by ``app.websocket_api.send`` if the provided ``connection_id`` is not
+connected anymore. In our case this doesn't really matter since we don't have
+anything tracking our connections. The error has a ``connection_id`` property
+that contains the offending connection id.
+
+.. code-block:: python
+
+   except WebsocketDisconnectedError as e:
+       pass  # Disconnected so we can't send the message back.
+
+
+To test out the echo server you can install ``websocket-client`` from PyPI::
+
+  pip install websocket-client
+
+
+After deploying the Chalice app the output will contain a URL for connecting
+to the websocket API labeled: ``- Websocket API URL:``. The
+``websocket-client`` package installs a command line tool called ``wsdump.py``
+which can be used to test websocket echo server::
+
+  $ wsdump.py wss://{websocket_api_id}.execute-api.us-west-2.amazonaws.com/api/
+  Press Ctrl+C to quit
+  > foo
+  < foo
+  > bar
+  < bar
+  > foo bar baz
+  < foo bar baz
+  >
+
+
+Every message sent to the server (lines that start with ``>``) result in a
+message sent to us (lines that start with ``<``) with the same content.
```

Comment:
It'd be helpful for this tutorial to follow the structure of the tutorial below, where you say when to write out the requirements files, and the command to run to deploy the echo server.